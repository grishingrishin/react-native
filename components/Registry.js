import React, { Component } from 'react';
import { TouchableWithoutFeedback } from 'react-native';
import * as ImagePicker from 'expo-image-picker';
import styled from 'styled-components';
import FadeInLoader from './FadeInLoader';
import FloatingLabelInput from './FloatingLabelInput';
import Question from './Question';

const Container = styled.ImageBackground`
  flex: 1;
  padding: 32px 0;
  justify-content: space-between;
  resize-mode: cover;
`;

const Header = styled.View`
  position: relative;
  margin: 32px 14px;
`;

const Title = styled.Text`
  text-transform: uppercase;
  text-align: center;
  font-size: 32px;
  font-family: Roboto-Thin;
  color: #fff;
`;

const Box = styled.View`
  padding: 0 14px;
`;

const Figure = styled.View`
  margin-bottom: 32px;
  align-items: center;
`;

const Ava = styled.Image`
  width: 180px;
  height: 180px;
  border-radius: 120px;
`;

const Upload = styled.Image`
  width: 150px;
  height: 150px;
`;

const Figcaption = styled.Text`
  margin-top: 14px;
  font-size: 28px;
  font-family: Roboto-Light;
  color: #fff;
`;

const Group = styled.View`
  margin-bottom: 14px;
`;

const Actions = styled.View``;

const Submit = styled.TouchableOpacity`
  align-items: center;
  height: 54px;
  background-color: #eb2054;
`;

const SubmitText = styled.Text`
  text-transform: uppercase;
  font-size: 18px;
  line-height: 54px;
  font-family: Roboto-Regular;
  color: #fff;
`;

class Register extends Component {
  constructor() {
    super();

    this.state = {
      username: '',
      email: '',
      password: '',
      passwordCheck: '',
      image: null,
    };
  }

  changeTextHandler = props => value => {
    return this.setState({
      [props]: value,
    });
  }

  pickImage = async () => {
    try {
      let result = await ImagePicker.launchImageLibraryAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.All,
        allowsEditing: true,
        aspect: [4, 3],
        quality: 1,
      });

      if (!result.cancelled) {
        this.setState({ image: result.uri });
      }

      return result;
    } catch (e) {
      console.log(e);
    }
  }

  render() {
    const { image } = this.state;

    return (
      <Container source={require('../assets/auth-bg.jpg')}>
        <Header>
          <Title>
            Create your account
          </Title>
        </Header>
        <FadeInLoader>
          <Box>
            <Figure>
              <TouchableWithoutFeedback onPress={this.pickImage}>
                { image ? <Ava source={{ uri: image }} /> : <Upload source={require('../assets/upload.png')} /> }
              </TouchableWithoutFeedback>
              <Figcaption>Upload your photo</Figcaption>
            </Figure>
            <Group>
              <FloatingLabelInput
                value={this.state.name}
                placeholder='Username'
                changeHandle={this.changeTextHandler('username')}
              />
            </Group>
            <Group>
              <FloatingLabelInput
                value={this.state.email}
                placeholder='Email address'
                changeHandle={this.changeTextHandler('email')}
              />
            </Group>
            <Group>
              <FloatingLabelInput
                value={this.state.password}
                placeholder='Password'
                changeHandle={this.changeTextHandler('password')}
              />
            </Group>
          </Box>
        </FadeInLoader>
        <Actions>
          <Submit activeOpacity={0.9}>
            <SubmitText>Sign up</SubmitText>
          </Submit>
          <Question 
            path='/'
            route='Sign in!'
            question='Have an account?'
          />
        </Actions>
      </Container>
    );
  }
}

export default Register;