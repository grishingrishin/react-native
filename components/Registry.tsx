import React, { Component } from 'react';
import { TouchableWithoutFeedback, Modal } from 'react-native';
import * as ImagePicker from 'expo-image-picker';
import styled from 'styled-components/native';
import FadeInLoader from './FadeInLoader';
import FloatingLabelInput from './FloatingLabelInput';
import Question from './Question';

const Container = styled.ImageBackground`
  flex: 1;
  padding: 32px 0;
  justify-content: space-between;
  resize-mode: cover;
`;

const Header = styled.View`
  position: relative;
  margin: 32px 14px;
`;

const Title = styled.Text`
  text-transform: uppercase;
  text-align: center;
  font-size: 32px;
  font-family: Roboto-Thin;
  color: #fff;
`;

const Box = styled.View`
  padding: 0 14px;
`;

const Figure = styled.View`
  margin-bottom: 32px;
  align-items: center;
`;

const Ava = styled.Image`
  width: 180px;
  height: 180px;
  border-radius: 120px;
`;

const Upload = styled.Image`
  width: 150px;
  height: 150px;
`;

const Figcaption = styled.Text`
  margin-top: 14px;
  font-size: 28px;
  font-family: Roboto-Light;
  color: #fff;
`;

const Group = styled.View`
  margin-bottom: 14px;
`;

const Actions = styled.View``;

const Submit = styled.TouchableOpacity`
  align-items: center;
  height: 54px;
  background-color: #eb2054;
`;

const SubmitText = styled.Text`
  text-transform: uppercase;
  font-size: 18px;
  line-height: 54px;
  font-family: Roboto-Regular;
  color: #fff;
`;

const ModalInner = styled.TouchableHighlight`
  flex: 1;
  justify-content: center;
  align-items: center;
  background-color: rgba(0,0,0,.5);
`;

const ModalCard = styled.View`
  width: 80%;
  padding: 18px 14px 0 14px;
  background-color: #fff;
`;

const ModalTitle = styled.Text`
  margin-bottom: 12px
  padding-bottom: 12px;
  font-size: 18px;
  line-height: 18px;
  font-family: Roboto-Medium;
  border-bottom-width: 1px;
  border-bottom-color: rgba(0, 0, 0, 0.1);
`;

const ModalText = styled.Text`
  margin-bottom: 18px;
  font-size: 16px;
`;

interface RegisterState {
  username: string,
  email: string,
  password: string,
  passwordCheck: string,
  image: string,
  pickImageModal: boolean,
}

class Register extends Component<{}, RegisterState> {
  constructor(props: object) {
    super(props);

    this.state = {
      username: '',
      email: '',
      password: '',
      passwordCheck: '',
      image: '',
      pickImageModal: false,
    };
  }

  handleUser = (username: string) => {
    return this.setState({
      username, 
    });
  }

  handleEmail = (email: string) => {
    return this.setState({
      email, 
    });
  }

  handlePassword = (password: string) => {
    return this.setState({
      password, 
    });
  }

  handlePasswordCheck = (passwordCheck: string) => {
    return this.setState({
      passwordCheck, 
    });
  }

  pickImage = () => {
    if (!this.state.pickImageModal) {
      return this.setState({
        pickImageModal: true
      });
    }
  }

  loadImageFromDevice = async () => {
    try {
      let result = await ImagePicker.launchImageLibraryAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.All,
        allowsEditing: true,
        aspect: [4, 3],
        quality: 1,
      });

      if (!result.cancelled) {
        return this.setState({
          image: result.uri,
          pickImageModal: false,
        });
      }

      return false;
    } catch (e) {
      console.log(e);
    }
  }

  loadImageFromCamera = async () => {
    try {
      let result = await ImagePicker.launchCameraAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.All,
        allowsEditing: true,
        aspect: [4, 3],
        quality: 1,
      });

      if (!result.cancelled) {
        return this.setState({
          image: result.uri,
          pickImageModal: false,
        });
      }

      return false;
    } catch (e) {
      console.log(e);
    }
  }

  removeImage = () => {
    return this.setState({
      image: '',
      pickImageModal: false,
    });
  }

  closedPickImageModal = () => {
    if (this.state.pickImageModal) {
      return this.setState({
        pickImageModal: false,
      });
    }
  }

  render() {
    const { image } = this.state;

    return (
      <Container source={require('../assets/auth-bg.jpg')}>
        <Modal transparent={true} visible={this.state.pickImageModal}>
          <ModalInner onPress={this.closedPickImageModal}>
            <ModalCard>
              <ModalTitle>Сменить фото профиля</ModalTitle>
              <TouchableWithoutFeedback onPress={this.loadImageFromDevice}>
                <ModalText>Загрузить с устройства</ModalText>
              </TouchableWithoutFeedback>
              <TouchableWithoutFeedback onPress={this.loadImageFromCamera}>
                <ModalText>Сделать снимок</ModalText>
              </TouchableWithoutFeedback>
              <TouchableWithoutFeedback onPress={this.removeImage}>
                <ModalText>Удалить фото</ModalText>
              </TouchableWithoutFeedback>
            </ModalCard>
          </ModalInner>
        </Modal>
        <Header>
          <Title>
            Create your account
          </Title>
        </Header>
        <FadeInLoader>
          <Box>
            <Figure>
              <TouchableWithoutFeedback onPress={this.pickImage}>
                { image ? <Ava source={{ uri: image }} /> : <Upload source={require('../assets/upload.png')} /> }
              </TouchableWithoutFeedback>
              <Figcaption>Upload your photo</Figcaption>
            </Figure>
            <Group>
              <FloatingLabelInput
                value={this.state.username}
                placeholder='Username'
                changeHandle={this.handleUser}
              />
            </Group>
            <Group>
              <FloatingLabelInput
                value={this.state.email}
                placeholder='Email address'
                changeHandle={this.handleEmail}
              />
            </Group>
            <Group>
              <FloatingLabelInput
                value={this.state.password}
                placeholder='Password'
                secureTextEntry={true}
                changeHandle={this.handlePassword}
              />
            </Group>
            <Group>
              <FloatingLabelInput
                value={this.state.passwordCheck}
                placeholder='Password check'
                secureTextEntry={true}
                changeHandle={this.handlePasswordCheck}
              />
            </Group>
          </Box>
        </FadeInLoader>
        <Actions>
          <Submit activeOpacity={0.9}>
            <SubmitText>Sign up</SubmitText>
          </Submit>
          <Question 
            path='/'
            route='Sign in!'
            question='Have an account?'
          />
        </Actions>
      </Container>
    );
  }
}

export default Register;