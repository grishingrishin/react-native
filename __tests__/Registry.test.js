import React from 'react';
import {shallow} from 'enzyme';
import {create} from 'react-test-renderer';
import Registry from '../components/Registry';

// Disable warning: `useNativeDriver` is not supported because the native animated module is missing. Falling back to JS-based animation.
// To resolve this, add `RCTAnimation` module to this app, or remove `useNativeDriver.
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
// Fix referenceError: You are trying to `import` a file after the Jest environment has been torn down.
jest.useFakeTimers();

describe('<Registry />', () => {
  describe('Jest tests', () => {
    it('renders correctly', () => {
      const tree = create(<Registry />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Enzyme tests', () => {
    it('allows us to loading new avatar', () => {
      const wrapper = shallow(<Registry />);

      wrapper.setState({image: '/images/'});
      expect(wrapper.state().image).toEqual('/images/');

      expect(wrapper.find('Registry__Ava')).toHaveLength(1);
    });

    it('allows us to simulate change event', () => {
      const wrapper = shallow(<Registry />);

      expect(wrapper.find('FloatingLabelInput')).toHaveLength(4);

      wrapper
        .find('FloatingLabelInput')
        .at(0)
        .props()
        .changeHandle('username@test');
      expect(wrapper.state().username).toEqual('username@test');

      wrapper
        .find('FloatingLabelInput')
        .at(1)
        .props()
        .changeHandle('email@test');
      expect(wrapper.state().email).toEqual('email@test');

      wrapper
        .find('FloatingLabelInput')
        .at(2)
        .props()
        .changeHandle('password@test');
      expect(wrapper.state().password).toEqual('password@test');

      wrapper
        .find('FloatingLabelInput')
        .at(3)
        .props()
        .changeHandle('password_check@test');
      expect(wrapper.state().passwordCheck).toEqual('password_check@test');
    });
  });
});