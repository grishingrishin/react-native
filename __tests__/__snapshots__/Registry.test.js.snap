// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registry /> Jest tests renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "space-between",
        "paddingBottom": 32,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 32,
        "resizeMode": "cover",
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../assets/auth-bg.jpg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "marginBottom": 32,
          "marginLeft": 14,
          "marginRight": 14,
          "marginTop": 32,
          "position": "relative",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontFamily": "Roboto-Thin",
            "fontSize": 32,
            "textAlign": "center",
            "textTransform": "uppercase",
          },
        ]
      }
    >
      Create your account
    </Text>
  </View>
  <View
    style={
      Object {
        "opacity": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "paddingBottom": 0,
            "paddingLeft": 14,
            "paddingRight": 14,
            "paddingTop": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "marginBottom": 32,
            },
          ]
        }
      >
        <Image
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          source={
            Object {
              "testUri": "../../../assets/upload.png",
            }
          }
          style={
            Array [
              Object {
                "height": 150,
                "width": 150,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
                "fontFamily": "Roboto-Light",
                "fontSize": 28,
                "marginTop": 14,
              },
            ]
          }
        >
          Upload your photo
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": 14,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "position": "relative",
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontSize": 22,
                "left": 14,
                "opacity": 1,
                "position": "absolute",
                "top": 18,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "Roboto-Light",
                  },
                ]
              }
            >
              Username
            </Text>
          </Text>
          <TextInput
            allowFontScaling={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            rejectResponderTermination={true}
            secureTextEntry={false}
            style={
              Array [
                Object {
                  "borderBottomColor": "rgba(211,200,230,0.5)",
                  "borderBottomWidth": 1,
                  "color": "#fff",
                  "fontFamily": "Roboto-Light",
                  "fontSize": 16,
                  "paddingBottom": 18,
                  "paddingLeft": 14,
                  "paddingRight": 14,
                  "paddingTop": 18,
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": 14,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "position": "relative",
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontSize": 22,
                "left": 14,
                "opacity": 1,
                "position": "absolute",
                "top": 18,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "Roboto-Light",
                  },
                ]
              }
            >
              Email address
            </Text>
          </Text>
          <TextInput
            allowFontScaling={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            rejectResponderTermination={true}
            secureTextEntry={false}
            style={
              Array [
                Object {
                  "borderBottomColor": "rgba(211,200,230,0.5)",
                  "borderBottomWidth": 1,
                  "color": "#fff",
                  "fontFamily": "Roboto-Light",
                  "fontSize": 16,
                  "paddingBottom": 18,
                  "paddingLeft": 14,
                  "paddingRight": 14,
                  "paddingTop": 18,
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": 14,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "position": "relative",
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontSize": 22,
                "left": 14,
                "opacity": 1,
                "position": "absolute",
                "top": 18,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "Roboto-Light",
                  },
                ]
              }
            >
              Password
            </Text>
          </Text>
          <TextInput
            allowFontScaling={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "borderBottomColor": "rgba(211,200,230,0.5)",
                  "borderBottomWidth": 1,
                  "color": "#fff",
                  "fontFamily": "Roboto-Light",
                  "fontSize": 16,
                  "paddingBottom": 18,
                  "paddingLeft": 14,
                  "paddingRight": 14,
                  "paddingTop": 18,
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": 14,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "position": "relative",
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontSize": 22,
                "left": 14,
                "opacity": 1,
                "position": "absolute",
                "top": 18,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "Roboto-Light",
                  },
                ]
              }
            >
              Password check
            </Text>
          </Text>
          <TextInput
            allowFontScaling={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "borderBottomColor": "rgba(211,200,230,0.5)",
                  "borderBottomWidth": 1,
                  "color": "#fff",
                  "fontFamily": "Roboto-Light",
                  "fontSize": 16,
                  "paddingBottom": 18,
                  "paddingLeft": 14,
                  "paddingRight": 14,
                  "paddingTop": 18,
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {},
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#eb2054",
          "height": 54,
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontFamily": "Roboto-Regular",
              "fontSize": 18,
              "lineHeight": 54,
              "textTransform": "uppercase",
            },
          ]
        }
      >
        Sign up
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 18,
            "textAlign": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontFamily": "Roboto-Thin",
              "fontSize": 12,
            },
          ]
        }
      >
        Have an account?
         
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
                "fontFamily": "Roboto-Regular",
                "fontSize": 12,
                "textDecorationColor": "black",
                "textDecorationLine": "underline",
                "textDecorationStyle": "solid",
              },
            ]
          }
        >
          Sign in!
        </Text>
      </View>
    </View>
  </View>
</View>
`;
